pkg/controllers/clusterresourceplacement/placement_controller.go

```diff
@@ -333,3 +339,116 @@ func (r *Reconciler) updatePlacementAppliedCondition(placement *fleetv1alpha1.Cl
 		}
 	}
 }
+
+// handleUpdate handles the create/update clusterResourcePlacement event.
+// It creates corresponding clusterPolicySnapshot and clusterResourceSnapshot if needed and updates the status based on
+// clusterPolicySnapshot status and work status.
+func (r *Reconciler) handleUpdate(ctx context.Context, crp *fleetv1.ClusterResourcePlacement) (ctrl.Result, error) {
+	crpKObj := klog.KObj(crp)
+	policyHash, err := generatePolicyHash(crp.Spec.Policy)
+	if err != nil {
+		klog.ErrorS(err, "Failed to generate policy hash of crp", "clusterResourcePlacement", crpKObj)
+		return ctrl.Result{}, err
+	}
+
+	clusterPolicySnapshotList := &fleetv1.ClusterPolicySnapShotList{}
+	if err := r.Client.List(ctx, clusterPolicySnapshotList, client.MatchingLabels{fleetv1.CRPTrackingLabel: crp.Name}); err != nil {
+		klog.ErrorS(err, "Failed to list clusterPolicySnapshots", "clusterResourcePlacement", crpKObj)
+		return ctrl.Result{}, err
+	}
+	latestPolicySnapshot, latestPolicySnapshotIndex, err := lookupLatestClusterPolicySnapshot(clusterPolicySnapshotList)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	if latestPolicySnapshot != nil && string(latestPolicySnapshot.Spec.PolicyHash) != policyHash {
+		// set the latest label to false first to make sure there is only one or none active policy snapshot
+		latestPolicySnapshot.Labels[fleetv1.IsLatestSnapshotLabel] = strconv.FormatBool(false)
+		if err := r.Client.Update(ctx, latestPolicySnapshot); err != nil {
+			klog.ErrorS(err, "Failed to set the isLatestSnapshot label to false", "clusterPolicySnapshot", klog.KObj(latestPolicySnapshot))
+			return ctrl.Result{}, err
+		}
+	}
+	if latestPolicySnapshot == nil || string(latestPolicySnapshot.Spec.PolicyHash) != policyHash {
+		// create a new policy snapshot
+		latestPolicySnapshotIndex++
+		latestPolicySnapshot = &fleetv1.ClusterPolicySnapshot{
+			ObjectMeta: metav1.ObjectMeta{
+				Name: fmt.Sprintf(fleetv1.PolicySnapshotNameFmt, crp.Name, latestPolicySnapshotIndex),
+				Labels: map[string]string{
+					fleetv1.CRPTrackingLabel:      crp.Name,
+					fleetv1.IsLatestSnapshotLabel: strconv.FormatBool(true),
+					fleetv1.PolicyIndexLabel:      strconv.Itoa(latestPolicySnapshotIndex),
+				},
+				OwnerReferences: []metav1.OwnerReference{ownerReference(crp)},
+			},
+			Spec: fleetv1.PolicySnapShotSpec{
+				Policy:     crp.Spec.Policy,
+				PolicyHash: []byte(policyHash),
+			},
+		}
+		if err := r.Client.Create(ctx, latestPolicySnapshot); err != nil {
+			klog.ErrorS(err, "Failed to create new clusterPolicySnapshot", "clusterPolicySnapshot", klog.KObj(latestPolicySnapshot))
+			return ctrl.Result{}, err
+		}
+	}
+	// create clusterResourceSnapshot
+	// update the status based on the latestPolicySnapshot status
+	// update the status based on the work
```

Comment:
What if the latestPolicy is not "active" but policy Hash is the same as the current? It seems that we won't do anything.

What if there are more than one "active" (not suppose to happen)? Do we want to safe guard that too?

Reply:
good catch, "What if the latestPolicy is not "active" but policy Hash is the same as the current? " should be handled and this corner case may happen. Will address it.

"What if there are more than one "active" (not suppose to happen)? Do we want to safe guard that too?"
Can we rely on the validating webhook & deny policy?